R/template.R

```diff
@@ -288,40 +348,58 @@ curve_metric_summarizer <- function(name,
     error_call = error_call
   )
 
-  if (!quo_is_null(case_weights)) {
+  if (quo_is_null(case_weights)) {
+    group_case_weights <- NULL
+  } else {
     case_weights <- yardstick_eval_select(
       expr = case_weights,
       data = data,
       arg = "case_weights",
       error_call = error_call
     )
-
-    case_weights <- expr(.data[[!!case_weights]])
   }
 
-  out <- dplyr::reframe(
-    data,
-    .metric = .env[["name"]],
-    .estimator = finalize_estimator(
-      .data[[truth]],
-      .env[["estimator"]],
-      .env[["name"]]
-    ),
-    .estimate = fn(
-      truth = .data[[truth]],
-      estimate = {
-        estimate <- dplyr::pick(tidyselect::all_of(.env[["estimate"]]))
-        prob_estimate_convert(estimate)
-      },
-      case_weights = !!case_weights,
-      na_rm = .env[["na_rm"]],
-      !!! spliceable_argument(estimator, "estimator"),
-      !!! spliceable_argument(event_level, "event_level"),
-      !!! fn_options
+  group_rows <- dplyr::group_rows(data)
+  groups <- vctrs::vec_chop(data, indices = group_rows)
+  out <- vector("list", length = length(groups))
+
+  for (i in seq_along(groups)) {
+    group <- groups[[i]]
+
+    group_truth <- group[[truth]]
+    group_estimate <- prob_estimate_convert(group[estimate])
+
+    if (is_string(case_weights)) {
+      group_case_weights <- group[[case_weights]]
+    }
+
+    elt_out <- c(
+      dplyr::group_keys(group),
+        list(
+        .metric = name,
+        .estimator = finalize_estimator(
+          group_truth,
+          estimator,
+          name
+        ),
+        .estimate = rlang::inject(
+          fn(
+            truth = group_truth,
+            estimate = group_estimate,
+            case_weights = group_case_weights,
+            na_rm = na_rm,
+            !!! spliceable_argument(estimator, "estimator"),
+            !!! spliceable_argument(event_level, "event_level"),
+            !!! fn_options
+          )
+        )
+      )
     )
-  )
 
-  dplyr::as_tibble(out)
+    out[[i]] <- tibble::as_tibble(elt_out)
```

Comment:
```suggestion
    out[[i]] <- tibble::new_tibble(elt_out)
```

typo or necessary?

Reply:
Necessary (or maybe bug in `new_tibble()`)

``` r
elt_out <- list(
  .metric = "roc_curve",
  .estimator = "binary",
  .estimate = tibble::tibble(
    .threshold = c(-Inf, 0.1, 0.5, Inf),
    specificity = c(0, 0.25, 0.7, 1),
    sensitivity = c(1, 0.8, 0.2, 0)
  )
)

tibble::as_tibble(elt_out)
#> # A tibble: 4 × 3
#>   .metric   .estimator .estimate$.threshold $specificity $sensitivity
#>   <chr>     <chr>                     <dbl>        <dbl>        <dbl>
#> 1 roc_curve binary                   -Inf           0             1  
#> 2 roc_curve binary                      0.1         0.25          0.8
#> 3 roc_curve binary                      0.5         0.7           0.2
#> 4 roc_curve binary                    Inf           1             0

tibble::new_tibble(elt_out)
#> Error in `[<-`:
#> ! Assigned data `map(.subset(x, unname), vectbl_set_names, NULL)` must
#>   be compatible with existing data.
#> ✖ Existing data has 1 row.
#> ✖ Element 3 of assigned data has 4 rows.
#> ℹ Row updates require a list value. Do you need `list()` or `as.list()`?
#> Caused by error in `vectbl_recycle_rhs_rows()`:
#> ! Can't recycle input of size 4 to size 1.
```

Reply:
Oh it is because `new_tibble()` assumes everything is the same length, and you need recycling to kick in first for the curve functions. You could use `new_tibble()` if you add this in before it:

```r
elt_out <- vctrs::vec_recycle_common(!!!elt_out)
```

It makes sense to do that for the curve summarizers. I'm not sure if it makes sense for `numeric_metric_summarizer()` and the other non-curve summarizers, because I think we expect that each group should return exactly 1 row there?

It is kind of nice that this forces an error if that is _not_ the case

Reply:
dealt with in [tidymodels/yardstick@`0c17e1b` (#405)](https://github.com/tidymodels/yardstick/pull/405/commits/0c17e1bbee9c66272c3956b95897ccb5c984be7f)